### PR TITLE
Intro data processor in sdgx

### PR DESCRIPTION
**Pull request by MooooCat**
## Description

In this PR, we introduced the data processor.

## Motivation and Context

Our data processor can transform and format a variety of data types to better preprocess the data sent to the model.

## How has this been tested?

This PR involves changes to multiple modules in sdgx, and testing will be done through unit tests (in tests/).

As of the time of this PR, we are working on unit tests.

## Types of changes

- [ ] Maintenance (no change in code, maintain the project's CI, docs, etc.)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: sdgx/data_models/metadata.py [hunk 1]
```diff
@@ -39,7 +39,7 @@ class Metadata(BaseModel):
     primary_keys is used to store single primary key or composite primary key
     """
 
-    column_list: Set[str] = set()
+    column_list: List[str] = []
```
Comment (Wh1isper) [inline sdgx/data_models/metadata.py:42]: I'm actually not too keen on changing column_list to list, I'm worried about duplicate processing happening once there are duplicate keys.

One possible way to make it a list is to use @property with setter and getter.

File: sdgx/data_processors/base.py [hunk 2]
```diff
@@ -1,16 +1,25 @@
...
     """
 
-    def fit(self, metadata: Metadata | None = None):
+    fitted = False
+
+    def check_fitted(self):
+        if not self.fitted:
+            raise SynthesizerProcessorError("Processor NOT fitted.")
+
+    def fit(self, metadata: Metadata | None = None, **kwargs: Dict[str, Any]):
```
Comment (Wh1isper) [inline sdgx/data_processors/base.py:22]:
```suggestion
    def fit(self, metadata: Metadata | None = None, **kwargs: Dict[str, Any]):
        self._fit(metadata, **kwargs)
        self.fitted = True
    def _fit(self, metadata: Metadata | None = None, **kwargs: Dict[str, Any]):
        raise NotImplementedError
```

This way the subclass won't forget to set the fitted.

File: sdgx/models/ml/single_table/ctgan.py [hunk 3]
```diff
@@ -205,11 +205,17 @@ def __init__(
         self._ndarry_loader = None
```
Comment (Wh1isper) [inline sdgx/models/ml/single_table/ctgan.py:206]:
```suggestion
self.data_dim = None
```

File: sdgx/synthesizer.py [hunk 4]
```diff
@@ -334,9 +345,16 @@ def sample(
...
         self.metadata = metadata  # Ensure update metadata
-        if metadata:
-            for d in self.data_processors:
-                d.fit(metadata)
+
+        # data_processors do not need to be fit again in the sampling stage
+        # if metadata:
+        #     for d in self.data_processors:
+        #         d.fit(metadata)
+
```
Comment (Wh1isper) [inline sdgx/synthesizer.py:353]: OK we can just remove them? Seems only need keep first line
```suggestion
        # data_processors do not need to be fit again in the sampling stage

```

File: sdgx/data_models/metadata.py (hunk 1 above)
Comment (MooooCat) [reply sdgx/data_models/metadata.py:42]: > I'm actually not too keen on changing column_list to list, I'm worried about duplicate processing happening once there are duplicate keys.
> 
> One possible way to make it a list is to use `@property` with setter and getter.

I believe it is necessary to maintain the order information of columns in metadata, and as you said `@property` should be a better solution.

**Review by Z712023 — APPROVED** [review]

**Pull request merged**